solidity-bindgen-macros/src/abi_gen.rs

```diff
@@ -0,0 +1,159 @@
+use crate::abi_json::{abi_from_json, Abi};
+use ethabi::param_type::{ParamType, Reader};
+use inflector::cases::snakecase::to_snake_case;
+use proc_macro2::{Ident, Span, TokenStream};
+use std::path::Path;
+
+pub fn abi_from_file(path: impl AsRef<Path>, span: Span) -> TokenStream {
+    let name = path
+        .as_ref()
+        .file_stem()
+        .unwrap()
+        .to_str()
+        .unwrap()
+        .to_owned();
+    let bytes = std::fs::read(path).unwrap();
+    let abis = abi_from_json(&bytes);
+    let abi_str = String::from_utf8(bytes).unwrap();
+
+    let struct_name = Ident::new(name.as_str(), span);
+
+    let fns = abis.iter().map(|abi| fn_from_abi(abi, span));
+
+    quote! {
+        pub struct #struct_name {
+            contract: ::solidity_bindgen::internal::ContractWrapper,
+        }
+
+        impl #struct_name {
+            pub fn new(address: ::web3::types::Address, url: &str, event_loop_handle: ::std::sync::Arc<::web3::transports::EventLoopHandle>) -> ::solidity_bindgen::internal::Result<Self> {
+                // Embed ABI into the program
+                let abi = #abi_str;
+
+                // Set up a wrapper so we can make calls
+                let contract = ::solidity_bindgen::internal::ContractWrapper::new(address, abi.as_bytes(), url, event_loop_handle)?;
+                Ok(Self {
+                    contract
+                })
+            }
+
+            #(#fns)*
+        }
+    }
+}
+
+fn param_token_type(param_type: ParamType) -> TokenStream {
+    match param_type {
+        ParamType::Address => quote! { ::web3::types::Address },
+        ParamType::Bytes => quote! { ::std::vec::Vec<u8> },
+        ParamType::Int(size) => match size {
+            256 => quote! { ::solidity_bindgen::internal::Unimplemented },
```

Comment:
Interesting. I would've expected an `I256` to exist. But I checked and that's not a thing (in rust-web3 at least). Odd but let's roll with it for now.

Reply:
It's also really weird in that they expose their own `U64` and `U128` types and don't explain why you might want to use them. 🤷